src/linux_native.rs

```diff
@@ -0,0 +1,670 @@
+            };
+
+        self.usage_page = usage_page;
+        self.initial = false;
+        self.value = &self.value[advanced..];
+        Some((usage_page, page))
+    }
+}
+
+// This comes from hidapi which apparently comes from Apple's implementation of
+// this
+fn next_hid_usage(
+    mut desc: &[u8],
+    initial: bool,
+    mut usage_page: u16,
+) -> Option<(u16, u16, usize)> {
+    let mut usage = None;
+    let mut usage_pair = None;
+    let mut advanced = 0_usize;
+
+    while !desc.is_empty() {
+        let key = desc[0];
+        let key_cmd = key & 0xfc;
+
+        let (data_len, key_size) = match hid_item_size(desc) {
+            Some(v) => v,
+            None => return None,
+        };
+
+        match key_cmd {
+            // Usage Page 6.2.2.7 (Global)
+            0x4 => {
+                usage_page = hid_report_bytes(desc, data_len) as u16;
+            }
+            // Usage 6.2.2.8 (Local)
+            0x8 => {
+                usage = Some(hid_report_bytes(desc, data_len) as u16);
+            }
+            // Collection 6.2.2.4 (Main)
+            0xa0 => {
+                // Usage is a Local Item, unset it
+                if let Some(u) = usage.take() {
+                    usage_pair = Some((usage_page, u))
+                }
+            }
+            // Input 6.2.2.4 (Main)
+		        0x80 |
+            // Output 6.2.2.4 (Main)
+		        0x90 |
+            // Feature 6.2.2.4 (Main)
+		        0xb0 |
+            // End Collection 6.2.2.4 (Main)
+		        0xc0  =>  {
+			          // Usage is a Local Item, unset it
+                usage.take();
+            }
+            _ => {}
+        }
+
+        advanced += data_len + key_size;
+        desc = &desc[(data_len + key_size)..];
+        if let Some((usage_page, usage)) = usage_pair {
+            return Some((usage_page, usage, advanced));
+        }
+    }
+
+    if let (true, Some(usage)) = (initial, usage) {
+        return Some((usage_page, usage, advanced));
+    }
+
+    None
+}
+
+/// Gets the size of the HID item at the given position
+///
+/// Returns data_len and key_size when successful
+fn hid_item_size(desc: &[u8]) -> Option<(usize, usize)> {
+    let key = desc[0];
+
+    // Long Item. Next byte contains the length of the data section.
+    if (key & 0xf0) == 0xf0 {
+        if !desc.is_empty() {
+            return Some((desc[1].into(), 3));
+        }
+
+        // Malformed report
+        return None;
+    }
+
+    // Short Item. Bottom two bits contains the size code
+    match key & 0x03 {
+        v @ 0 | v @ 1 | v @ 2 => Some((v.into(), 1)),
+        3 => Some((4, 1)),
+        _ => unreachable!(), // & 0x03 means this can't happen
+    }
+}
+
+/// Get the bytes from a HID report descriptor
+///
+/// Must only be called with `num_bytes` 0, 1, 2 or 4.
+fn hid_report_bytes(desc: &[u8], num_bytes: usize) -> u32 {
+    if num_bytes >= desc.len() {
+        return 0;
+    }
+
+    match num_bytes {
```

Comment:
Probably makes more sense to not match but rather just do  this:

```rust
let mut bytes: [0u8; 4] = [0; 4];
bytes[..num_bytes].copy_from_slice(desc[..num_bytes]);
u32::from_le_bytes(bytes)
```

Also the num_bytes param may be unneeded? Why not just use the slice len

Reply:
I think that does work. It's clearer now that the endian conversion isn't done by hand. The `num_bytes` is the actual amount of data that we should read, while `desc` is the whole rest of the descriptor. Now, we should be able able to call this function with the equivalent of `&desc[..num_bytes]` in the parent, but I'm not sure where it's less awkward.